pagefaultgames/pokerogue
# Refactor Import Functions to Resolve Circular Dependency Issues in Tests

## What are the changes?
I've refactored the code by converting all anonymous functions executed during imports into named functions.
These named functions are now executed in the loading scene.

This change prevents crashes during tests caused by circular dependencies, which occur when anonymous functions are imported in an unpredictable order.

This refactoring is crucial because, for example, the biomes module crashes if the pokemon-evolution module isn't imported first, and the pokemon-evolution module crashes if the pokemon-forms module isn't imported first, and so on.

## Why am I doing these changes?
This change is introduced to fix the circular dependency issues causing crashes during imports in tests. This improvement comes from the need to write better tests and ensure stable imports. The current state with anonymous functions executed at import leads to unpredictable behavior and crashes.

## What did change?
Explicitly, the changes involve wrapping the previously anonymous functions into named functions. For example:

### Before:
```typescript
(function () {
  const statKeys = Object.keys(displayStats);

  for (const key of statKeys) {
    if (typeof displayStats[key] === "string") {
      let label = displayStats[key] as string;
      let hidden = false;
      if (label endsWith("?")) {
        label = label.slice(0, -1);
        hidden = true;
      }
      displayStats[key] = {
        label: label,
        sourceFunc: gameData => gameData.gameStats[key].toString(),
        hidden: hidden
      };
    } else if (displayStats[key] === null) {
      displayStats[key] = {
        sourceFunc: gameData => gameData.gameStats[key].toString()
      };
    }
    if (!(displayStats[key] as DisplayStat).label) {
      const splittableKey = key.replace(/([a-z]{2,})([A-Z]{1}(?:[^A-Z]|$))/g, "$1_$2");
      (displayStats[key] as DisplayStat).label = Utils.toReadableString(`${splittableKey[0].toUpperCase()}${splittableKey.slice(1)}`);
    }
  }
})();
```

### After:
```typescript
export function initStatsKeys() {
  const statKeys = Object.keys(displayStats);

  for (const key of statKeys) {
    if (typeof displayStats[key] === "string") {
      let label = displayStats[key] as string;
      let hidden = false;
      if (label endsWith("?")) {
        label = label.slice(0, -1);
        hidden = true.
      }
      displayStats[key] = {
        label: label,
        sourceFunc: gameData => gameData.gameStats[key].toString(),
        hidden: hidden
      };
    } else if (displayStats[key] === null) {
      displayStats[key] = {
        sourceFunc: gameData => gameData.gameStats[key].toString()
      };
    }
    if (!(displayStats[key] as DisplayStat).label) {
      const splittableKey = key.replace(/([a-z]{2,})([A-Z]{1}(?:[^A-Z]|$))/g, "$1_$2");
      (displayStats[key] as DisplayStat).label = Utils.toReadableString(`${splittableKey[0].toUpperCase()}${splittableKey.slice(1)}`);
    }
  }
}
```

This change was applied to the following files:
- `src/ui/game-stats-ui-handler.ts` -> `initStatsKeys()`
- `src/data/pokemon-evolutions.ts` -> `initPokemonPrevolutions()`
- `src/data/biomes.ts` -> `initBiomes()`
- `src/data/egg-moves.ts` -> `initEggMoves()`
- `src/data/pokemon-forms.ts` -> `initPokemonForms()`

With these changes, we can now write tests like:
```typescript
it("should create an enemypokemon with specified moveset", () => {
  const ennemyPokemon = new PokemonData({
    species: Species.MEWTWO,
    level: 100,
    moveset: [
      new PokemonMove(Moves.ACID),
      new PokemonMove(Moves.ACROBATICS),
      new PokemonMove(Moves.FOCUS_ENERGY),
    ]
  });
  expect(ennemyPokemon.moveset[0].moveId).toBe(Moves.ACID);
  expect(ennemyPokemon.moveset[1].moveId).toBe(Moves.ACROBATICS);
  expect(ennemyPokemon.moveset[2].moveId).toBe(Moves.FOCUS_ENERGY);
});
```

Without this fix, errors occur when importing. With this fix, no more errors are present.

### Screenshots/Videos
Without this fix:
![image](https://github.com/pagefaultgames/pokerogue/assets/44787002/3253ff68-7d36-41ac-9045-9a02394dc405)

With this fix
![image](https://github.com/pagefaultgames/pokerogue/assets/44787002/132f2edf-36e6-4820-b5a7-1ae7f56ce30e)

## How to test the changes?
To test these changes:
1. Check out the branch.
2. Import the mentionned files in the tests.
3. Run some tests to ensure they pass without spitting unknown errors.

## Checklist
